File: python/experiment.py
"""Experiment LiTS."""
import os
import sys
import argparse
from database import LiTSdb
from scan import LiTSscan
from preprocessor import LiTSpreprocessor
from rough_detector import LiTSroughDetector
from tumor_detector import LiTStumorDetector
import time


def main():
    """Main function for the liver tumor detection training and testing."""
    # Input arguments parsing
    # _______________________________________________________________________ #
    parser = argparse.ArgumentParser(description=''
                                     'Liver tumor segmentation experiment')
    parser.add_argument('-i', dest='db_path', required=True,
                        help='Path to the directory where folders '
                             '"Training Batch 1" and '
                             '"Training Batch 2" are placed')
    args = parser.parse_args()
    if not os.path.exists(args.db_path):
        print("\nInput database path does not exist!\n")
        sys.exit(1)
    if not os.path.exists(os.path.join(args.db_path, 'Training Batch 1')):
        print("\nTraining Batch 1 is not in the input database directory!\n")
        sys.exit(1)
    if not os.path.exists(os.path.join(args.db_path, 'Training Batch 2')):
        print("\nTraining Batch 2 is not in the input database direcory!\n")
    # _______________________________________________________________________ #

    # Database object creation and data split creation.
    # _______________________________________________________________________ #
    db = LiTSdb(args.db_path)
    subjects = db.get_subjects_names()
    training, validation, testing = db.get_data_split(subjects, 1)
    # _______________________________________________________________________ #

    # Data loading, preprocessing and liver bounding box detection
    # _______________________________________________________________________ #
    preprocess = LiTSpreprocessor()
    rough_detector = LiTSroughDetector()
    scans = {}
    time_start = time.time()
    for s in subjects[0:]:
        print("s:", s)
        if s not in scans:
            scans[s] = LiTSscan(s)

        # Loading CT volume, segmentation labels and voxel size,
        # volume intensity preprocessing and patient orientation
        # ___________________________________________________________________ #
        scans[s].load_volume(db.get_volume_path(s))
        scans[s].load_segmentation(db.get_segmentation_path(s))
        scans[s].load_info(db.get_volume_path(s))
        preprocess.preprocess(scans[s], s)
        rough_detector.detect_liver_bbox(scans[s])

    time_end = time.time()
    print("time elapsed:", time_end - time_start)
    # _______________________________________________________________________ #

    # Segmentation training and testing
    # _______________________________________________________________________ #
    tumor_detector = LiTStumorDetector()

    tumor_detector.train(scans, training, validation)
    tumor_detector.test(scans, testing)

if __name__ == '__main__':
    main()

File: c++/src/experiment.cpp
#include <iostream>
#include <stdlib.h>
#include <string.h>
#include <boost/filesystem.hpp>
#include "lits_db/lits_database.h"
#include "lits_scan/lits_scan.h"


int main(int argc, char **argv)
{
	if(argc != 2)
	{
		std::cout<<"Invalid number of input arguments."<<std::endl;
		std::cout<<"First argument: directory path where the Train Batch 1 and Train Batch 2 are placed."<<std::endl;
		exit(EXIT_FAILURE);
	}

	if(!strcmp(argv[1], "-h") or !strcmp(argv[1], "-help") or !strcmp(argv[1], "help") or !strcmp(argv[1], "h"))
	{
		std::cout<<"comparison:"<<strcmp(argv[1], "-h")<<std::endl;
		std::cout<<"First argument: directory path where the Train Batch 1 and Train Batch 2 are placed."<<std::endl;
		exit(EXIT_SUCCESS);
	}

	boost::filesystem::path db_path(argv[1]);

	if(boost::filesystem::is_directory(db_path))
	{
		std::string db_path_str(argv[1]);
		boost::filesystem::path db_path_batch((db_path_str + "/Training Batch 1").c_str());
		if(!boost::filesystem::is_directory(db_path_batch))
		{
			std::cout<<"Training Batch 1 does not exists."<<std::endl;
			exit(EXIT_FAILURE);
		}

		db_path_batch = db_path_str + "/Training Batch 2";
		if(!boost::filesystem::is_directory(db_path_batch))
		{
			std::cout<<"Training Batch 2 does not exists."<<std::endl;
			exit(EXIT_FAILURE);
		}

	}
	else
	{
		std::cout<<"Invalid database path!"<<std::endl;
		exit(EXIT_FAILURE);
	}

	LiTS_db db(db_path.string());
	db.load_subject_names();
	db.data_split(0, 1);

	for(unsigned int i = 0; i < 5; i++)
	{
		std::string volume_path;
		std::string segmentation_path;
		std::string scan_name;

		scan_name = db.get_train_scan_name(i);
		db.get_scan_paths(scan_name, volume_path, segmentation_path);

		LiTS_scan ls(volume_path, segmentation_path);
		ls.load_volume();
		ls.load_segmentation();
		ls.load_info();
	}

	return 0;
}
